providersymcrypt.go

```diff
@@ -0,0 +1,374 @@
+	_SymCryptBlobTypeSha1State       = _SymCryptBlobTypeHashState + 4
+	_SymCryptBlobTypeSha256State     = _SymCryptBlobTypeHashState + 5
+	_SymCryptBlobTypeSha384State     = _SymCryptBlobTypeHashState + 6
+	_SymCryptBlobTypeSha512State     = _SymCryptBlobTypeHashState + 7
+	_SymCryptBlobTypeSha3_256State   = _SymCryptBlobTypeHashState + 8
+	_SymCryptBlobTypeSha3_384State   = _SymCryptBlobTypeHashState + 9
+	_SymCryptBlobTypeSha3_512State   = _SymCryptBlobTypeHashState + 10
+	_SymCryptBlobTypeSha224State     = _SymCryptBlobTypeHashState + 11
+	_SymCryptBlobTypeSha512_224State = _SymCryptBlobTypeHashState + 12
+	_SymCryptBlobTypeSha512_256State = _SymCryptBlobTypeHashState + 13
+	_SymCryptBlobTypeSha3_224State   = _SymCryptBlobTypeHashState + 14
+
+	_SYMCRYPT_MD5_STATE_EXPORT_SIZE    = uint32(unsafe.Sizeof(_SYMCRYPT_MD5_STATE_EXPORT_BLOB{}))
+	_SYMCRYPT_SHA1_STATE_EXPORT_SIZE   = uint32(unsafe.Sizeof(_SYMCRYPT_SHA1_STATE_EXPORT_BLOB{}))
+	_SYMCRYPT_SHA256_STATE_EXPORT_SIZE = uint32(unsafe.Sizeof(_SYMCRYPT_SHA256_STATE_EXPORT_BLOB{}))
+	_SYMCRYPT_SHA512_STATE_EXPORT_SIZE = uint32(unsafe.Sizeof(_SYMCRYPT_SHA512_STATE_EXPORT_BLOB{}))
+)
+
+type _SYMCRYPT_BLOB_HEADER struct {
+	magic uint32
+	size  uint32
+	_type uint32
+}
+
+type _SYMCRYPT_BLOB_TRAILER struct {
+	checksum [8]uint8
+}
+
+// _UINT64 is a 64-bit unsigned integer, stored in native endianess.
+// It is used to represent a SymCrypt UINT64 type without making the
+// parent struct 8-byte aligned, given that the Windows ABI makes
+// the struct 4-byte aligned.
+type _UINT64 [2]uint32
+
+func newUINT64(v uint64) _UINT64 {
+	var u _UINT64
+	if isBigEndian {
+		u[0], u[1] = uint32(v>>32), uint32(v)
+	} else {
+		u[0], u[1] = uint32(v), uint32(v>>32)
+	}
+	return u
+}
+
+func (u *_UINT64) uint64() uint64 {
+	if isBigEndian {
+		return uint64(u[0])<<32 | (uint64(u[1]))
+	}
+	return uint64(u[0]) | (uint64(u[1]) << 32)
+}
+
+// symCryptAppendBinary appends the binary representation of a SymCrypt state
+// to the given destination slice.
+func symCryptAppendBinary(dst, chain, buffer []byte, blength _UINT64) []byte {
+	length := blength.uint64()
+	var nx uint64
+	if len(buffer) <= 64 {
+		nx = length & 0x3f
+	} else {
+		nx = length & 0x7f
+	}
+	dst = append(dst, chain...)
+	dst = append(dst, buffer[:nx]...)
+	dst = append(dst, make([]byte, len(buffer)-int(nx))...)
+	dst = appendUint64(dst, length)
+	return dst
+}
+
+// symCryptUnmarshalBinary unmarshals the binary representation of a SymCrypt state
+// from the given source slice. It returns the length of the data.
+func symCryptUnmarshalBinary(d []byte, chain, buffer []byte) _UINT64 {
+	copy(chain[:], d)
+	d = d[len(chain):]
+	copy(buffer[:], d)
+	d = d[len(buffer):]
+	_, length := consumeUint64(d)
+	return newUINT64(length)
+}
+
+// swapEndianessInt32 swaps the endianness of the given byte slice
+// in place. It assumes the slice is a backup of a 32-bit integer array.
+func swapEndianessInt32(d []uint8) {
```

Comment:
Calling it `Int32` rather than `Uint32` might be confusing later? AFAIK it doesn't matter, but in that case, I figure it 
 might as well match the rest of the file.